rancher/system-agent-installer-rke2
# Partially revert run.ps1 to handle continuous reboot of RKE2 service

### Issue: 

### Problem:
Changes made to the `run.ps1` file in  https://github.com/rancher/system-agent-installer-rke2/pull/40 unintentionally resulted in the inability to upgrade the kubernetes version of windows worker nodes.

### Solution:

Remove sleep periods before and after restarting the RKE2 windows service. These changes result in the worker being properly upgraded, as well maintains the fix made for https://github.com/rancher/system-agent-installer-rke2/pull/40.


This issue was encountered due to the RKE2 windows service restarting itself in spite of a graceful shutdown. The RKE2 service is configured to restart upon a failure every 30 seconds, and the prior update of this script also paused for 30 seconds with the intention of allowing the service some time to free any resources in use. However, this resulted in the RKE2 service transitioning into a running state at the same time as the script attempts to upgrade the version. 

The fatal error encountered with the previous version of this script which prevented proper upgrades can be seen in the RKE2 service logs, indicating that an instance of RKE2 is already running and using the containerd shim exe on the host. 
```
6/20/2023 5:23:28 PM {failed to extract runtime image: open C:\var\lib\rancher\rke2\data\v1.24.11-rke2r1-windows-amd64-def9d7224936\bin\containerd-shim-runhcs-v1.exe: The process cannot access the file because it is being used by another process.}                          
```

[The Microsoft documentation states](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/ns-winsvc-service_failure_actionsa) that the service is only expected to automatically restart if the service terminates without transitioning to the `SERVICE_STOPPED` state (i.e. a non graceful shutdown). `run.ps1` initiates a graceful shutdown of the service through the use of the `Stop-Service` command (this is acknowledged by RKE2 via the following log message: `6/20/2023 12:55:03 PM {Windows Service is shutting down gracefully}`).

Looking at the RKE2 source code,  [RKE2's service implementation takes this requirement into account and properly signals that state during a graceful shutdown](https://github.com/rancher/rke2/blob/02f58981a09e61f578c306c4e417aa336bf170eb/pkg/windows/service_windows.go#L42-L46). I'm still investigating why windows is restarting this service after graceful shut downs, and I've removed aspects of this script which depend on that expected behavior. In the mean time, forcibly killing the process provides a path forward and resolves the upgrading issues as well as the issues #40 aimed to address. Let me know if there are concerns with this approach, I haven't seen any impact to functionality in my testing. 


### Testing: 
1. Upgrade Kubernetes Cluster with a windows worker node 
 + Create a custom cluster using a Kubernetes version which includes this change
 + Provision a Linux node with all roles and wait for it to become available 
 + Provision a windows worker node and register it to the custom cluster (This was tested using 2019 and 2022). 
 + Initiate an upgrade via the Rancher UI to another Kubernetes version also including these changes. 
 + Observe the cluster properly upgrade and become available 

2. Graceful reboot of the node does not result in node disconnection 

+ Create a custom cluster using a Kubernetes version which includes these changes 
+ Register am all roles linux node to the cluster and wait for it to become available 
+ Within vSphere, provision a windows VM and configure it as needed 
     1. The 'Containers' feature needs to be enabled 
     2. VMWare tools need to be installed onto the VM 
+ Register windows worker node to existing custom cluster and wait for it to become available 
+ Within the vSphere UI, gracefully restart the host 
+ Observe the worker node temporarily become unavailable, and after some time properly reconnect to the cluster. 
+ Power cycle the VM several more times in the vSphere UI, observe the node reconnect consistently
